### PR TITLE
[ROCM] activated bfloat16 tests for rocm platform

### PR DESCRIPTION
**Pull request by pemeliya**
In this PR, we added a one central function **IsBF16NotSupportedInOps** in tensorflow/core/kernels/gpu_utils.cc to decide whether bfloat16 datatypes are directly supported on the target acrhitecture or not. 
Previously, the similar checks were spread around several files. 
It will facilitate the integration of bfloat16 support to ROCM: currently bfloat16-to-float conversion is always enabled on ROCM.

Besides, we also enabled bfloat16 support in a number of subtests.
Finally, rewrote tensorflow/core/kernels/batch_norm_op_test.cc in a generic form to be able to test it for different number types.

### CODE REVIEW

File: tensorflow/core/kernels/gpu_utils.h [hunk 1]
```diff
@@ -42,6 +42,10 @@ class AutotuneResult;
 
 namespace tensorflow {
 
+// returns true if bfloat16 is not directly supported in Ops and inputs shall be
+// casted to floats to perform the computations and then back
+bool IsBF16NotSupportedInOps(se::Stream *stream);
```
Comment (cantonios) [inline tensorflow/core/kernels/gpu_utils.h:47]: Let's flip this so it's `IsBF16Supported` to avoid double-negatives when checking for support.

File: tensorflow/core/kernels/batch_norm_op_test.cc [hunk 2]
```diff
@@ -29,60 +29,58 @@ limitations under the License.
...
+
+TYPED_TEST_P(BatchNormOpTest, Simple) {
+  this->run_me();
 }
 
+REGISTER_TYPED_TEST_SUITE_P(BatchNormOpTest,
+                            Simple);
+
+// TODO(ezhulenev): Add support for more data types.
+using DataTypes = ::testing::Types<float, Eigen::half>;//, Eigen::bfloat16>;
```
Comment (cantonios) [inline tensorflow/core/kernels/batch_norm_op_test.cc:80]: Are you planning to add bfloat16 here?  Why is that commented?

File: tensorflow/core/kernels/gpu_utils.cc [hunk 3]
```diff
@@ -37,6 +37,20 @@ using xla::AutotuningLog;
...
+  if (!stream) {
+    return true;  // no stream: don't know whether it's supported
+  }
+#if GOOGLE_CUDA  
+// Performant bfloat16 operations are supported for Ampere+ GPUs. For
+// pre-Ampere GPUs, we cast inputs to float and outputs back to bfloat16.
+  return !stream->GetCudaComputeCapability().IsAtLeast(
+          se::CudaComputeCapability::AMPERE);  
+#elif TENSORFLOW_USE_ROCM
+  return true;  // so far, we return true meaning that the conversion to float is needed
```
Comment (cantonios) [inline tensorflow/core/kernels/gpu_utils.cc:50]: nit: comments should start with a capital letter and have proper punctuation (i.e. period at the end).

**Review by cantonios — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/batch_norm_op_test.cc (hunk 2 above)
Comment (pemeliya) [reply tensorflow/core/kernels/batch_norm_op_test.cc:80]: I have rechecked it, the point is, there is no registered BatchNormWithGlobalNormalization OpKernel for bfloat16 number type: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/batch_norm_op.cc#L197-L203 
I will remove it then..

**Comment by pemeliya** [comment]
Requested changes are implemented, I wonder if there are any futher issues with this PR ? Thanks!

**Review by cantonios — APPROVED** [review]

File: tensorflow/core/kernels/fused_batch_norm_op.cc [hunk 4]
```diff
@@ -34,6 +34,7 @@ limitations under the License.
 #include "tensorflow/core/framework/register_types.h"
 #include "tensorflow/core/framework/tensor.h"
 #include "tensorflow/core/framework/tensor_types.h"
+#include "tensorflow/core/kernels/gpu_utils.h"
```
Comment (cantonios) [inline tensorflow/core/kernels/fused_batch_norm_op.cc:37]: You need to add a bazel dep to include this header.

**Review by cantonios — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/fused_batch_norm_op.cc (hunk 4 above)
Comment (pemeliya) [reply tensorflow/core/kernels/fused_batch_norm_op.cc:37]: indeed, I missed a couple of places which caused cpu build brake, thanks

**Review by cantonios — APPROVED** [review]

**Review by cantonios — CHANGES_REQUESTED** [review]

**Review by cantonios — APPROVED** [review]
